logstash-core/src/main/java/org/logstash/config/ir/CompiledPipeline.java

```diff
@@ -0,0 +1,385 @@
+        final Map<String, RubyIntegration.Filter> res =
+            new HashMap<>(filterPlugins.size(), 1.0F);
+        for (final PluginVertex plugin : filterPlugins) {
+            final String ident = plugin.getId();
+            if (!res.containsKey(ident)) {
+                res.put(ident, buildFilter(plugin));
+            }
+        }
+        return res;
+    }
+
+    /**
+     * Sets up all Ruby inputs learnt from {@link PipelineIR}.
+     */
+    private Collection<IRubyObject> setupInputs() {
+        final Collection<PluginVertex> vertices = pipelineIR.getInputPluginVertices();
+        final Collection<IRubyObject> nodes = new HashSet<>(vertices.size());
+        vertices.forEach(v -> {
+            final PluginDefinition def = v.getPluginDefinition();
+            final SourceWithMetadata source = v.getSourceWithMetadata();
+            nodes.add(pipeline.buildInput(
+                RubyUtil.RUBY.newString(def.getName()), RubyUtil.RUBY.newFixnum(source.getLine()),
+                RubyUtil.RUBY.newFixnum(source.getColumn()), convertArgs(def)
+            ));
+        });
+        return nodes;
+    }
+
+    /**
+     * Converts plugin arguments from the format provided by {@link PipelineIR} into coercible
+     * Ruby types.
+     * @param def PluginDefinition as provided by {@link PipelineIR}
+     * @return RubyHash of plugin arguments as understood by {@link RubyIntegration.Pipeline}
+     * methods
+     */
+    private RubyHash convertArgs(final PluginDefinition def) {
+        final RubyHash converted = RubyHash.newHash(RubyUtil.RUBY);
+        for (final Map.Entry<String, Object> entry : def.getArguments().entrySet()) {
+            final Object value = entry.getValue();
+            final String key = entry.getKey();
+            final Object toput;
+            if (value instanceof PluginStatement) {
+                final PluginDefinition codec = ((PluginStatement) value).getPluginDefinition();
+                toput = pipeline.buildCodec(
+                    RubyUtil.RUBY.newString(codec.getName()),
+                    Rubyfier.deep(RubyUtil.RUBY, codec.getArguments())
+                );
+            } else {
+                toput = value;
+            }
+            converted.put(key, toput);
+        }
+        return converted;
+    }
+
+    /**
+     * Compiles a {@link RubyIntegration.Filter} from a given {@link PluginVertex}.
+     * @param vertex Filter {@link PluginVertex}
+     * @return Compiled {@link RubyIntegration.Filter}
+     */
+    private RubyIntegration.Filter buildFilter(final PluginVertex vertex) {
+        final PluginDefinition def = vertex.getPluginDefinition();
+        final SourceWithMetadata source = vertex.getSourceWithMetadata();
+        return pipeline.buildFilter(
+            RubyUtil.RUBY.newString(def.getName()), RubyUtil.RUBY.newFixnum(source.getLine()),
+            RubyUtil.RUBY.newFixnum(source.getColumn()), convertArgs(def)
+        );
+    }
```

Comment:
Maybe use the same pattern for the input / output?

Reply:
@ph same answer here, only way to do that is to pull Ruby String constants over to Java.
I think the only other approach really is to use a more sophisticated wrapper for all of this, but  to me it seems like it's just noise since all of this is kinda in-flux right now => time for adding wrappers and such for the Ruby end is probably better spent moving more of the logic to Java instead?